common/src/abi.rs

```diff
@@ -175,6 +176,27 @@ pub fn format_token(param: &Token) -> String {
     }
 }
 
+/// Formats an integer as a string together with its exponential notation.
+pub fn format_uint_with_exponential_hint(num: &U256) -> String {
+    if num.lt(&U256::from(10_000)) {
+        return num.to_string()
+    }
+
+    let stringified = num.to_string();
+    let exponent = stringified.len() - 1;
+    let mut base = stringified.chars().take(4).collect::<String>();
+
+    base = base.trim_end_matches('0').to_string();
+
+    if base.len() > 1 {
+        base.insert(1, '.');
+    }
```

Comment:
I haven't fully understood this,

what does this do?

Reply:
to build the exponential notation from the stringified number:

1. take the power of 10 (len - 1)
2. take the first 4 digits because I want to show up to 3 decimal digits (1 digit, the `.` and 3 digits after it)
3. trim the ending zeroes (in cases such as `1.000e18` I opted to show `1e18` instead of `1.000e18` for brevity since we already have the fully expanded number on the left anyway
4. insert the `.` character only if there are decimals left (for instance in the above example show `1e18` and not `1.e18`)

I think this implementation is quite fast since it's just manipulating a string rather than having to deal with potential floating point precision limitations, especially on very large integers which are very common in this case. Let me know if it makes sense!